front: redesign of the list of trains on the results page

theocrsb opened the pull request:
Mockup: https://www.sketch.com/s/2fb064cd-c483-46ac-bd30-72705fd63e75/a/el9lZmq

Closes #8601

theocrsb committed:
front: redesign of the list of trains on the results page

Co-authored-by: Simon Ser <contact@emersion.fr>
Signed-off-by: theocrsb <theo_crosbie@yahoo.fr>
Signed-off-by: Simon Ser <contact@emersion.fr>

theocrsb commented:
https://www.sketch.com/s/2fb064cd-c483-46ac-bd30-72705fd63e75/p/27329D7B-C7CF-48C4-835A-1BA94627AF52/canvas

kmer2016 commented on front/src/styles/scss/applications/operationalStudies/_scenario.scss:
```diff
@@ -133,351 +155,463 @@
+              pointer-events: auto;
+            }
+
+            .scenario-timetable-train-button {
+              margin-left: 0.5rem;
+              &:hover {
+                background-color: var(--white);
+                color: var(--primary80);
               }
             }
+            .scenario-timetable-train-button-1 {
+              margin-left: 0;
+            }
+          }
```
```suggestion
            .scenario-timetable-train-button {
              margin-left: 0.5rem;
              &:hover {
                background-color: var(--white);
                color: var(--primary80);
              }
              
              &:first-child {
               margin-left: 0;
              }
            }
          }
```

theocrsb replied:
Done

kmer2016 commented on front/src/modules/trainschedule/components/Timetable/TimetableTrainCard.tsx:
```diff
@@ -150,173 +138,131 @@ const TimetableTrainCard = ({
+      </div>
+      <div className="scenario-timetable-train-buttons">
+        <button
+          className="scenario-timetable-train-button-1"
```
You can select the first child using the css pseudo class `first-child` instead of attach a specif class. You can see the suggestion below ;)
```suggestion
          className="scenario-timetable-train-button"
```

theocrsb replied:
Thanks for the tip, it's done

kmer2016 reviewed (changes requested):
"Great work! Not much to say on the code side, just a small CSS suggestion. However, I have a few questions regarding the behavior:

- It seems like a third scrollbar has appeared.
![image](https://github.com/user-attachments/assets/260aa6ad-fc60-4e3e-bc32-aae264fd6240)

- When clicking the checkbox on a TrainCard, the corresponding train gets selected, and the curves and output tables are updated with that train's info. Previously, the train was selected by clicking on the TrainCard itself, not the checkbox.
- In a new scenario, the 'No train' checkbox is always checked and cannot be unchecked.
- There's no longer any distinction for trains on which path projection has been performed.
- The hover effect on the first action button doesn't turn white like the others."

theocrsb commented:
> "Great work! Not much to say on the code side, just a small CSS suggestion. However, I have a few questions regarding the behavior:
> 
> * It seems like a third scrollbar has appeared.
>   ![image](https://private-user-images.githubusercontent.com/19575568/367214217-260aa6ad-fc60-4e3e-bc32-aae264fd6240.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYyMTczNTMsIm5iZiI6MTcyNjIxNzA1MywicGF0aCI6Ii8xOTU3NTU2OC8zNjcyMTQyMTctMjYwYWE2YWQtZmM2MC00ZTNlLWJjMzItYWFlMjY0ZmQ2MjQwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEzVDA4NDQxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM1ZjE4NDAzMTQ5NzE5ODVkZjcwMzRkZTkxODdjMGMxZTU4NjRmYzA5NWE1YWM0MTcyMDg5ZTRlNDU1MmI1NGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.x5xs9ik7fol_3kenG6dnB0roLqEIc8S51UFXyjsdOys)
> * When clicking the checkbox on a TrainCard, the corresponding train gets selected, and the curves and output tables are updated with that train's info. Previously, the train was selected by clicking on the TrainCard itself, not the checkbox.
> * In a new scenario, the 'No train' checkbox is always checked and cannot be unchecked.
> * There's no longer any distinction for trains on which path projection has been performed.
> * The hover effect on the first action button doesn't turn white like the others."

This is done for the hover and scroll bar.
When there is no train, it is normal for the box to be checked. 
For trains on which a path projection has been carried out, @thibautsailly will be able to give us a new display.
We'll need to solve the checkbox problem.

theocrsb commented:
I pushed a solution for the checkbox problem

kmer2016 reviewed (approved):
LGTM !

emersion committed:
front: drop scenario timetable background color

The collapsed timetable header has a background which doesn't match
the rest of the app.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion commented:
Differences from the mockup:

- "Select path" icon doesn't exist in ui-icons, so old one is kept as-is.
- Arrow icon on the right of "show more details" is missing, but that's [by design](https://github.com/OpenRailAssociation/osrd/issues/8601#issuecomment-2320459295).
- "Invalid rolling stock" text is displayed aligned to the right to leave more room for the train name.
- Checkmark in checked checkbox is not at the same position as the mockup - seems like a ui-core bug.
- Buttons don't look the same as the mockup - seems like a ui-core bug.

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/TimetableToolbar.tsx:
```diff
@@ -140,100 +139,111 @@ const TimetableToolbar = ({
-                className="mr-2"
-                checked={selectedTrainIds.length === trainSchedulesWithDetails.length}
+      <div
+        className={cx('scenario-timetable-toolbar justify-content-between', {
```
Since we have a class anyway, maybe we could insert `justify-content` inside it ?

```suggestion
        className={cx('scenario-timetable-toolbar', {
```

theocrsb replied:
`justify-content-between` is not enabled in the browser, so I'll remove it.

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/TimetableTrainCard.tsx:
```diff
@@ -150,173 +134,132 @@ const TimetableTrainCard = ({
-                <div className="scenario-timetable-train-arrival" title={train.arrivalTime}>
-                  {formatDateTimeWithoutYear(train.arrivalTime)}
-                </div>
+              {!train.rollingStock && t('timetable.invalidRollingStock')}
```
We no longer detail why the train is invalid ? (@thibautsailly)

thibautsailly replied:
> We no longer detail why the train is invalid ? (@thibautsailly)

It wasn't an acceptance criteria, so I didn't know about it. I can figure something out if need be, let me know.

Castavo replied:
After discussion with @thibautsailly the "invalidity reason" is displayed in plain text, in the place where the train would be. (like for "MAtériel invalide" in the mockup)

theocrsb replied:
Done

SharglutDev commented on front/src/styles/scss/_variables.scss:
```diff
@@ -32,6 +32,18 @@ $colors: (
   'coolgray5': #b9b9b9,
   'coolgray3': #d7d7d7,
   'coolgray1': #f2f2f2,
+  'grey60': #5c5955,
+  'ambiantB5': #faf9f5,
+  'ambiantB10': #f7f6ee,
+  'ambiantB15': #f2f0e4,
+  'selection20': #fff2b3,
+  'warning30': #eaa72b,
+  'warning60': #7d521e,
+  'error60': #d91c1c,
+  'error80': #6b0000,
+  'error5': #ffeeed,
+  'primary60': #1844ef,
+  'primary80': #1f0f96,
```
Maybe we could sort the variables and separate them from the "old" colors

```suggestion

  // OSRD Design System
  'grey60': #5c5955,
  'ambiantB5': #faf9f5,
  'ambiantB10': #f7f6ee,
  'ambiantB15': #f2f0e4,
  'selection20': #fff2b3,
  'warning30': #eaa72b,
  'warning60': #7d521e,
  'error60': #d91c1c,
  'error80': #6b0000,
  'error5': #ffeeed,
  'primary60': #1844ef,
  'primary80': #1f0f96,
```

theocrsb replied:
You're right, it's done

emersion commented on front/public/locales/fr/operationalStudies/scenario.json:
```diff
@@ -50,7 +50,12 @@
+      "pathfinding_not_found": "Chemin introuvable",
+      "pathfinding_failed": "Calcul interrompu",
+      "simulation_failed": "Simulation impossible",
+      "rolling_stock_not_found": "MR non trouvé"
```
Is "MR" explicit enough? Is it a commonly used acronym?

theocrsb replied:
MR has been validated by @thibautsailly

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/TimetableToolbar.tsx:
```diff
@@ -140,100 +139,111 @@ const TimetableToolbar = ({
+      >
+        <div className="toolbar-header">
+          {trainSchedules.length === 0 ? (
+            <div className="no-train">
```
Mockup : 
![Capture d’écran 2024-09-24 à 16 30 26](https://github.com/user-attachments/assets/4f8baa08-543b-40af-94d1-113ed94117a9)

PR : 
![Capture d’écran 2024-09-24 à 16 30 18](https://github.com/user-attachments/assets/7d5ee645-bdf4-4b66-861c-829702b72716)

Color is wrong (you need to add `grey80` variable) and it seems like the alignment is not correct

theocrsb replied:
Done

SharglutDev replied:
Seems like you added the grey80 in variables but it's not used and alignment i still not correct

SharglutDev replied:
Also if you don't use the grey80 after all, you can remove it

theocrsb replied:
Done

SharglutDev replied:
Ok for the "no train" but when there are trains

theocrsb replied:
I've just opened a [bug ticket](https://github.com/OpenRailAssociation/osrd-ui/issues/552) in the `ui` repository. When the bug is solved, the problem should also be solved.

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/TimetableTrainCard.tsx:
```diff
@@ -150,173 +134,132 @@ const TimetableTrainCard = ({
-                <div className="scenario-timetable-train-arrival" title={train.arrivalTime}>
-                  {formatDateTimeWithoutYear(train.arrivalTime)}
-                </div>
+              {train.invalidReason && t(`timetable.invalid.${train.invalidReason}`)}
```
![Capture d’écran 2024-09-24 à 16 33 44](https://github.com/user-attachments/assets/43978e25-4c71-43cd-a8ea-0626960a845b)

The img should not be there if the train is invalid I think and the text color/weight seems wrong

theocrsb replied:
Done

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/FilterPanel.tsx:
```diff
@@ -63,81 +64,95 @@ const FilterPanel = ({
-          </div>
+      </div>
+      <div className="compositions-code">
+        <label htmlFor="composition-tag-filter">{t('timetable.compositionCodes')}</label>
```
Mockup : 
![Capture d’écran 2024-09-24 à 16 38 03](https://github.com/user-attachments/assets/a6c27ed2-9ecd-4824-91b7-7c4d41a59a2b)

PR : 
![Capture d’écran 2024-09-24 à 16 39 05](https://github.com/user-attachments/assets/c72d4575-2479-4384-b28d-ab827cf355b0)

Seems like the margin between label and select is too big (and the text is not the same ?)

theocrsb replied:
Done

SharglutDev commented on front/src/styles/scss/_variables.scss:
```diff
@@ -24,14 +24,28 @@ $colors: (
+  'ambiantB10': #f7f6ee,
+  'ambiantB15': #f2f0e4,
   'coolgray1': #f2f2f2,
+  'coolgray3': #d7d7d7,
```
`coolgray`'s come from the old design system, they should stay in the other section

theocrsb replied:
Done

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_scenario.scss:
```diff
@@ -133,351 +152,440 @@
+          .label {
+            margin-left: 0.5rem;
+            font-size: 1rem;
+            font-weight: 400;
```
Don't think it's needed

```suggestion
```

theocrsb replied:
Done

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/TimetableTrainCard.tsx:
```diff
@@ -150,173 +134,132 @@ const TimetableTrainCard = ({
-    <div className="scenario-timetable-train-with-right-bar">
+    <div
+      data-testid="scenario-timetable-train"
+      className={cx('scenario-timetable-train', {
```
Mockup : 
![Capture d’écran 2024-09-24 à 16 44 21](https://github.com/user-attachments/assets/584d2d6f-d0fc-40af-9bae-9d940da7d51c)

PR : 
![Capture d’écran 2024-09-24 à 16 44 40](https://github.com/user-attachments/assets/4c606c07-400a-465a-9411-7e581057f34a)

In the mockup, it seems like the background of cards is brighter than the parent background but not in the PR (backgrounds almost seem the same ?)

theocrsb replied:
Done

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/TimetableTrainCard.tsx:
```diff
@@ -150,173 +134,132 @@ const TimetableTrainCard = ({
+                {t('timetable.stopsCount', { count: train.stopsCount })}
+              </span>
+              <span className="more-info-item">{train.pathLength}</span>
+              <span className="more-info-item" data-testid="allowance-energy-consumed">
```
![Capture d’écran 2024-09-24 à 16 44 40](https://github.com/user-attachments/assets/e2142325-971c-48a5-95fd-f5b6d3fcf073)

Spacing between details seems wrong

theocrsb replied:
Done

SharglutDev replied:
Ok for the energy consumed but the duration goes to next line in large screens

theocrsb replied:
Done

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/TimetableTrainCard.tsx:
```diff
@@ -147,173 +131,131 @@ const TimetableTrainCard = ({
-                <div className="scenario-timetable-train-arrival" title={train.arrivalTime}>
-                  {formatDateTimeWithoutYear(train.arrivalTime)}
-                </div>
+              {train.invalidReason && t(`timetable.invalid.${train.invalidReason}`)}
```
Seems like on large screens, the position of the invalid reason is too much on left

![image](https://github.com/user-attachments/assets/c65615bd-110c-4b96-9c50-7ee36311eb15)

theocrsb replied:
Done
@thibautsailly sent me this screen
mockup: ![image](https://github.com/user-attachments/assets/745f4769-eb19-4ed3-aadc-eb786ad63096)
app: 
<img width="432" alt="image" src="https://github.com/user-attachments/assets/c8f4f11a-d761-4f22-ba8c-5da4d513c39e">

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/Timetable.tsx:
![image](https://github.com/user-attachments/assets/abe80f04-a5a0-433f-bf01-e8d76d3282d2)

When I have multiple trains, when scrolling down, I don't see the filter panel and the header anymore. Should they be fixed ? @thibautsailly 

For a long list of trains, the user would have to scroll to the top all over to change a filter or to hide/show details

thibautsailly replied:
Yes they should be fixed. Sorry for missing this use case and not specifying it. And the list container should be as tall as possible, but we should work on this when we do the whole page.

theocrsb replied:
Done

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/TimetableToolbar.tsx:
```diff
@@ -140,100 +139,111 @@ const TimetableToolbar = ({
+        <div className="toolbar-header">
+          {trainSchedules.length === 0 ? (
+            <div className="no-train">
+              <Checkbox small label={t('timetable.noTrain')} />
```
I think the UI of the checkbox is the read-only one. If the label of the component doesn't match the position of the mockup, the text shouldn't be inserted in label

theocrsb replied:
Done

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/TimetableToolbar.tsx:
```diff
@@ -140,100 +139,111 @@ const TimetableToolbar = ({
+                }
                 onChange={() => toggleAllTrainsSelecton()}
               />
+              <div className="mx-2 text-base" data-testid="train-count">
```
Why do we need text-base ?

theocrsb replied:
We don't need it

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_scenario.scss:
```diff
@@ -133,351 +152,437 @@
+          align-items: center;
+
+          .label {
+            margin-left: 0.625rem;
```
0.5rem was the right margin I think

theocrsb replied:
Done

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_scenario.scss:
```diff
@@ -133,351 +152,437 @@
+                text-align: left;
+                line-height: 1.25rem;
+                padding: 0.125rem 0.5rem;
+                margin: 0 0.75rem 0.5rem 0;
```
You don't need this margin, just put a `gap: 0.75rem` on the parent (gap is 12px on mockup)

```suggestion
```

theocrsb replied:
Done

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_scenario.scss:
```diff
@@ -133,351 +152,437 @@
+          }
+
+          .compositions-code {
+            padding: 0.625rem 0.813rem 1rem 1rem;
```
I don't think the spacing between the bottom of this component and the end of the filter panel should be handled partly by the padding bottom of this component. 
Mockup needs a 23px spacing, it should be directly by the parent with `.filter-panel` class.

theocrsb replied:
I pushed a solution where I adapted to the padding already existing on reusable components.

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/Timetable.tsx:
```diff
@@ -131,45 +123,43 @@ const Timetable = ({
+              removeTrains={removeAndUnselectTrains}
             />
           ))}
+        <div
+          className={cx('bottom-timetables-trains', {
+            'empty-list': trainSchedulesWithDetails.length === 0,
+          })}
+        />
```
I think we don't need this div. It just corresponds to the bottom of the `.scenario-timetable-trains` div. It should work with just a padding-bottom and the special `.empty-list` class when no train

theocrsb replied:
You're right, I can't remember why I did it but it's not useful. I've just changed it.

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_scenario.scss:
```diff
@@ -133,351 +152,428 @@
+
+      .scenario-timetable-trains {
+        font-family: 'IBM Plex Sans';
+        background: linear-gradient(
+          to top,
+          var(--ambiantB15) calc(0.938rem),
+          rgba(0, 0, 0, 0.25) calc(1rem),
+          var(--ambiantB15) calc(1.063rem)
+        );
```
I can't find any trace of this gradient in the mockup. It seems to be just a regular ambiantB15.

theocrsb replied:
Done

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_scenario.scss:
```diff
@@ -133,351 +152,428 @@
+          background-color: var(--ambiantB15);
+        }
+        .empty-list {
+          height: 7.5rem;
```
This should be `7rem;` and probably a padding-bottom instead of height when you will remove the bottom div

theocrsb replied:
Done

theocrsb committed:
front: redesign filter panel

Co-authored-by: Simon Ser <contact@emersion.fr>
Signed-off-by: theocrsb <theo_crosbie@yahoo.fr>
Signed-off-by: Simon Ser <contact@emersion.fr>

SharglutDev reviewed (approved):
Lgtm and tested, great job :) Thank you @emersion for participating as well !
